javascript/ql/src/semmle/javascript/security/dataflow/ZipSlip.qll

```diff
@@ -33,6 +33,33 @@ module ZipSlip {
     }
   }
 
+  /**
+   * Holds if `node1` flows to `node2` in one step by virtue of
+   * `node2` being of the form `.pipe(node1)`. The reason this flow
+   * exists is that `.pipe` returns its argument to make chained
+   * stream operations work.
+   */
+  predicate pipeStep(DataFlow::Node node1, DataFlow::MethodCallNode node2) {
+      node2.getMethodName() = "pipe" and
+      node1 = node2.getArgument(0)
+  }
+
+  /**
+   * Holds if `node1` flows to `node2` in one step including the assumption that
+   * `x` flows to `.pipe(x)`
+   */
+  predicate stepsThroughPipe(DataFlow::Node node1, DataFlow::Node node2) {
+    DataFlow::localFlowStep(node1, node2) or pipeStep(node1, node2)
+  }
+
+  /**
+   * Holds if `node1` flows to `node2` including the assumption that
+   * `x` flows to `.pipe(x)`
+   */
+  predicate flowsThroughPipe(DataFlow::Node node1, DataFlow::Node node2) {
+    stepsThroughPipe*(node1, node2)
```

Comment:
Unfortunately this relation is huge. `DataFlow::localFlowStep` is very large, which means `stepsThroughPipe` is also large, and we're only interested in a very small subset of this graph. Materializing this graph is a bit expensive, let alone computing its transitive closure.

In cases like this I think it's better to use explicit recursion: (untested)
```
DataFlow::SourceNode parsedStream() {
  result = DataFlow::moduleImport("unzip").getAMemberCall("Parse")
  or
  exists (DataFlow::MethodCallNode pipe |
    pipe.getMethodCall() = "pipe" and
    pipe.getArgument(0).getALocalSource() = parsedStream() and
    result = pipe
  )
)
```
You can also write this with the transitive closure operator if you prefer, but it should not be over a relation that contains `localFlowStep`.

Reply:
Ah, great, thanks, this really helps me fill out my mental model of how the evaluator is likely to deal with these kinds of patterns. I was trying to formulate a question of the form "ok so I think what I want is to customize local data flow the same way `Configuration`s let you customize interprocedural data flow, and this is all I could think to arrive at given I couldn't find any extension points" but the performance considerations you're describing make sense.